src/charm.py

```diff
@@ -0,0 +1,293 @@
+
+        aar_snap = snap.SnapCache()["aar"]
+        aar_snap.restart()
+        self.unit.status = ops.ActiveStatus()
+
+    def _register_ams_client(self, client_certificate: str, mode: str):
+        try:
+            client_certificate = client_certificate.replace("\\n", "\n").strip('"')
+            cmd = ["/snap/bin/aar", "trust", "add"]
+            if mode == "publisher":
+                cmd.append("--publisher")
+            check_output(cmd, stderr=STDOUT, input=client_certificate.encode("utf-8"))
+        except CalledProcessError as ex:
+            logger.error(f"failed to add client to aar: {ex.output}")
+            raise
+
+    def _snap_installed(self, snap_name) -> bool:
+        snap_client = snap.SnapClient()
+        snaps = snap_client.get_installed_snaps()
+        for installed_snap in snaps:
+            if installed_snap["name"] == snap_name:
+                return True
+        return False
+
+    def _get_ip_for_interface(self, interface):
+        """Return the ip address associated to the given interface."""
+        addresses = netifaces.ifaddresses(interface)
+        if netifaces.AF_INET not in addresses or "addr" not in addresses[netifaces.AF_INET][0]:
+            raise Exception("No IP associated to requested device")
+
+        return addresses[netifaces.AF_INET][0]["addr"]
+
+    @property
+    def public_ip(self) -> str:
+        """Public address of the unit."""
+        public_iface = self.config["public_interface"]
+        if public_iface:
+            public_address = self._get_ip_for_interface(public_iface)
+            if public_address:
+                return public_address
+            logger.warning(
+                "Could not obtain a valid IP for the configured public_interface. \
+                Using the default one"
+            )
+        return self.model.get_binding("juju-info").network.ingress_address.exploded
+
+    @property
+    def private_ip(self) -> str:
+        """Private address of the unit."""
+        return self.model.get_binding("juju-info").network.bind_address.exploded
+
+    def _create_cert_files_if_needed(self, cert_path, key_path):
+        if os.path.exists(cert_path) and os.path.exists(key_path):
+            return
+
+        cert, key = self._generate_selfsigned_cert(self.public_ip, self.public_ip, self.private_ip)
+
+        cert_base_path = os.path.dirname(cert_path)
+        if not os.path.exists(cert_base_path):
+            os.makedirs(cert_base_path, mode=0o0700)
+
+        key_base_path = os.path.dirname(key_path)
+        if not os.path.exists(key_base_path):
+            os.makedirs(key_base_path, mode=0o0700)
+
+        with open(os.open(cert_path, os.O_CREAT | os.O_WRONLY, 0o600), "w") as f:
+            f.write(str(cert, "UTF-8"))
+
+        with open(os.open(key_path, os.O_CREAT | os.O_WRONLY, 0o600), "w") as f:
+            f.write(str(key, "UTF-8"))
+
+    def _generate_selfsigned_cert(self, hostname, public_ip, private_ip):
```

Comment:
There is https://github.com/canonical/self-signed-certificates-operator but to remain backward compatible this is fine to do. However there seems to be existing logic in the interface (see https://github.com/canonical/self-signed-certificates-operator/blob/main/lib/charms/tls_certificates_interface/v2/tls_certificates.py#L681). Is that maybe something we can reuse as it's exported to public users of the lib?